### ISSUE TITLE
Route 'stops' are randomly moving after loading a save

### ISSUE THREAD

**Issue by VictorPhilipp** [CitiesSkylinesMods/TMPE#244]
(from @holmas)

Every time after loading a saved game several public transport routes are damaged. This is because extra stops have been inserted at strange places. So far I have been able to repair the route by clearing the additional stops. But this time the additional stop is inaccessible. The attachment contains 2 screenshots and the saved game. The first screenshot shows the broken monorail route (the black one) and the second one shows the absolutely strange place of the additional stop.

By the way: thank you soo much for this incredible work

Original issue: https://github.com/VictorPhilipp/Cities-Skylines-Traffic-Manager-President-Edition/issues/244

**Comment by VictorPhilipp** [comment]
I also encountered a similar issue. As far as I remember I could rule out TMPE as the cause. Do you use the Improved Public Transport 2 mod? It might causing this issue but I am not 100% sure.

**Comment by VictorPhilipp** [comment]
Nodes were randomly assigned to other public transport lines. This happened not only while saving/loading but also during the game. However, I was unable to find the root cause and could find a way to reproduce the behavior - it only happend by chance.

I also inspected whether TMPE is able to modify or create new transport line nodes and if transport line assignments are touched. This is not the case, thus I don't believe that TMPE is causing this issue.

**Comment by holmas** [comment]
I'm using three mods only: Fine Road Anarchy, Building Themes and - of course - TMPE. In the mean time this town becomes unplayable, due to tons of NullReferenceExeceptions. What a pity ... For a moment I thought it could be a path finder issue. Thats why I reported this here.

**Comment by VictorPhilipp** [comment]
```
Simulation error: Object reference not set to an instance of an object
  at Building.CalculateGroupData (UInt16 buildingID, Int32 layer, System.Int32& vertexCount, System.Int32& triangleCount, System.Int32& objectCount, .VertexArrays& vertexArrays) [0x00000] in <filename unknown>:0 
  at BuildingManager.CalculateGroupData (Int32 groupX, Int32 groupZ, Int32 layer, System.Int32& vertexCount, System.Int32& triangleCount, System.Int32& objectCount, .VertexArrays& vertexArrays) [0x00000] in <filename unknown>:0 
  at RenderManager.Managers_CalculateGroupData (Int32 groupX, Int32 groupZ, Int32 layer, System.Int32& vertexCount, System.Int32& triangleCount, System.Int32& objectCount, .VertexArrays& vertexArrays) [0x00000] in <filename unknown>:0 
  at RenderGroup.UpdateMeshData (.MeshLayer layer, UnityEngine.Vector3& min, UnityEngine.Vector3& max) [0x00000] in <filename unknown>:0 
  at RenderGroup.UpdateMeshData () [0x00000] in <filename unknown>:0 
  at RenderManager.SimulationStepImpl (Int32 subStep) [0x00000] in <filename unknown>:0 
  at SimulationManagerBase`2[Manager,Properties].SimulationStep (Int32 subStep) [0x00000] in <filename unknown>:0 
  at RenderManager.ISimulationManager.SimulationStep (Int32 subStep) [0x00000] in <filename unknown>:0 
  at SimulationManager.SimulationStep () [0x00000] in <filename unknown>:0 
  at SimulationManager.SimulationThread () [0x00000] in <filename unknown>:0   [Core]
```

I can't load the savegame, I guess some asset might be broken.

**Comment by VictorPhilipp** [comment]
An asset is missing:
```
Unknown prefab: HM_5Lane+Parking_RA+Monorail.5-Lane roundabout w. Monorail_Data  [Serialization]
 
(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)

Unknown prefab: HM_5Lane+Parking_Roundabout.5-Lane-Roundabout_Data  [Serialization]
```

**Comment by VictorPhilipp** [comment]
Fine Road Anarchy is deprecated.

**Comment by holmas** [comment]
I’ve tried my own roads - download zipped assets https://www.sovoma.de/downloads/priv/Assets.zip <https://www.sovoma.de/downloads/priv/Assets.zip>. Furthermore, I’ve rejected Fine Road Anarchy, but the problem with the broken lines still exists.

I think the Mass Transit DLC can to be a suspect. Above-average often these "ghost stops“ will be created on a monorail path when the route is a bus route and vice versa.

**Comment by aubergine10** [comment]
Fine Road Anarchy is the up-to-date one IIRC. It's the Advanced Road Anarchy that seems to cause probs.

From a mod I'm working on...

```
// working mods:
            option.Add(802066100, 1); // Fine Road Anarchy by SamSamTS
            option.Add(651322972, 2); // Fine Road Tool by SamSamTS

// deprecated and may cause bugs:
            deprecated.Add(1362508329, 1); // Fine Road Anarchy 2018
            deprecated.Add(433567230, 3);  // Advanced Road Anarchy
            deprecated.Add(1436255148, 1); // Fine Road Anarchy (Chinese)

// known to be game breaking:
            obsolete.Add(448434637, 2); // * Enhanced Road Heights
            obsolete.Add(418556522, 3); // * Road Anarchy
            obsolete.Add(954034590, 3); // * Road Anarchy V2
            obsolete.Add(553184329, 1); // * Sharp Junction Angles
```

**Comment by holmas** [comment]
I'm using nothing of them and even Fine Road Anarchy were not used in this specific saved game.

**Comment by VictorPhilipp** [comment]
Unfortunately I was unable to reproduce your problem in ~8 hours of gameplay with multiple save and load actions. The issue seems to occur so randomly and seldomly that it seems impossible to reproduce/track it down to a specific cause.

However, I found one suspicious things during my analysis: TMPE's custom path manager and the `StartPathFind` methods are currently initialized on level loading (after deserialization). But at that time they could already have been used for path calculation. Maybe a stock path becomes inconsistent because of some kind of (timing?) issue during the replacement of the stock manager / deploying detours?
A possible problematic execution path could like this:
1. `TransportLineAI.UpdatePath` detects that a public transport segment must be recalculated. It calls `TransportLineAI.StartPathFind` (stock method)
2. `StartPathFind` executes a stock path-find. Path-finding starts.
3. `PathManager`/`PathFind` are detoured while path-finding takes place. Queue items are replaced by custom structs. Linking between different path units is now carried out via `PathUnitQueueItem.nextPathUnitId` (instead of `BufferItem.m_nextPathUnit`).
![grafik](https://user-images.githubusercontent.com/4333935/55267239-6200f700-5281-11e9-9ece-03ccd4c32184.png)
4. Since `PathFind` methods are detoured during PF execution those linked lists (or in general the `PathFind` instance state) become inconsistent and path-finding starts (randomly) assigning path units of other calculated paths to a public transport path (for example a regular road segment now becomes part of a monorail path).
5. `TransportLineAI.CheckNodePosition` is fed with a path position originating from a corrupt path unit and derives an erroneous node position from it.
6. `NetManager.MoveNode` moves the public transport stop to the erroneous position.

Possible fix: Replacing the stock path manager and detouring `StartPathFind` methods could be performend before data is deserialized (which is executed before `OnLevelLoaded` is called).

**Comment by holmas** [comment]
There must be a big difference between our computers, due to I'm still fighting with these broken lines - and by no means seldomly. I certainly don't dare to build a new road anymore, after that I have to fix several lines for 10 minutes. Next, I will remove steam and all games completely (only 3 :D) und rebuild it from scratch.

**Comment by VictorPhilipp** [comment]
Very interesting, do you have any **disabled** mods in your content manager? Maybe one of them is causing this. Even if they are disabled they still can interfere with the game.

**Comment by holmas** [comment]
Oh, I have. There is Fine Road Anarchy and First Person Camera. Both of them are deactivated only. I just banished them to the Big Nothingness. Let's see what happen ...

**Comment by aubergine10** [comment]
I use Fine Road Anarchy and First Person Camera ([the 'Updated' one](https://steamcommunity.com/sharedfiles/filedetails/?id=650805785&searchtext=first+person+camera)) and haven't noticed issues like that.

There's definitely reports that "Advanced Road Anarchy" mod was (at least at some point in the past) creating ghost nodes / segments. A recent update to that mod sought to mitigate the issue with a feature to scan for and clean up such breakages. I assume such breakages would persist in save game even if the Advanced Road Anarchy mod was uninstalled.

I did had weird PT stop 'relocation issues' in the past, which I suspect were caused by IPT2 (not applicable in this case as holmas doesn't use that mod).

> Above-average often these "ghost stops“ will be created on a monorail path when the route is a bus route and vice versa.

Could you explain that a bit more? Does it happen if there's a bus route near a monorail path?

**Comment by holmas** [comment]
Currently, I made a backup of my steam installation, so I can (want) not load any saved game. I believe one of the saved games I introcduced in this issue discussion should have exactly this situation. Near the IT stack is a monorail station which has a single broken line marker. This "ghost" stop comes from the black bus route. The black bus line runs far away from this monorail station in the southwest. Near the monorail station runs the light green line, which is not broken here.  The reverse case is happen too several times, always with the blue monorail line. In this case, a small piece of a monorail line will be displayed on a road with stops at both ends. These "magic stops" are not accessible in either way. Deleting the whole line is the only fix I've found so far. Unfortunately, I do not have a saved game with this curious line run.

**Comment by VictorPhilipp** [comment]
I saw the stop you mention, it's from monorail line M3. You can access it by building a monorail station such that it aligns with the dot. Then build a monorail track and connect your existing stations with the temporary station. When the problem icon goes away remove the temporary station and monorail tracks.

**Comment by holmas** [comment]
I'm already used to constantly rebuilding my lines ;). In Germany we say "Practice make perfect". So, I think, I'm perfect lol ...

**Comment by VictorPhilipp** [comment]
You mean "Übung macht den Meister" ;-)

**Comment by holmas** [comment]
Genau!

**Comment by aubergine10** [comment]
I recall that the base game repositions stops if roads/rails are edited. For example, in the case of busses, if you delete a road segment that had a bus stop the game will automatically place the stop on a nearby road. With train stations it is different in that it will leave the stop where it is (although last time I tried my game was heavily modded).

@VictorPhilipp Is it worth checking in to the code that handles stop location deletion in the base game? Also I suspect that mods like IPT2 and TLM might hook in to such code as they add additional logic relating to stop and would need to know of any changes. Is it possible that TMPE is causing some sort of issue when the stop relocation code kicks in (maybe race condition where TMPE pathfinding hasn't updated at point where stop relocation code kicks in)?

**Comment by aubergine10** [comment]
Not sure if it's related, but some Steam users discussing issues with routes breaking (getting gaps in them that they can't fix in the usual manner eg. dragging a stop and the route sorts itself out): https://steamcommunity.com/app/255710/discussions/0/1768134097428159187/

I can ask for saves / logs if required?

**Comment by socalledwolverine** [comment]
Hi, 
I'm experiencing the same problem. I'm posting my savegame:

[LakefieldTMPE.zip](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/3104555/LakefieldTMPE.zip)

**Comment by aubergine10** [comment]
Just had potentially another 2 users with this problem, asked both for save game.

Also wondering if it's possible this could be caused by broken node problem described in #277 ?

For anyone struggling to remove bus routes once they become broken, see this youtube vid by Steam user OneJasonBradly: https://www.youtube.com/watch?v=V40FuyQywGs

**Comment by aubergine10** [comment]
Another report on LABS page:

> **Yangxuzhu:** Hello ! so i have a strange issue on my big city, it happen sometimes, (fortunately) that i notice a broken public transport line appear with no reason, ( once on bus line, twice on metro line, once on train line) the line broke and stops appearing in weird places (like on the opposite map where there is not even a track or station for the train stop for ex) therefore the line is so mess up, i cant even fix it (i can drag stop or remove but i cannot reconnect the line) so i have no choice but to delete the line and redo it again, I suspect the issue may come from TMPE ( i use the LABS) because i have this strange issue since i install this mod, but well i am not sure.. so i want to know if anyone have encountered this problem as well, apart from this, great mod.

**Comment by aubergine10** [comment]
A user reported that they removed all other mods (except TM:PE) and the problem stopped. Then they added Network Extensions 2 and it started happening again.

https://steamcommunity.com/workshop/filedetails/discussion/1637663252/2266942917238024778/?tscn=1588212505

**Comment by krzychu124** [comment]
Closing due to inactivity. Reopen if necessary

**Comment by aTadnMad** [comment]
oh, this is still an issue.  i have a large city, and i am pretty fed up with my bus lines being broken, repairing them, only to have them break again, or others be broken.  like others have said, sometimes i find bus stops in the middle of the ocean!  i'm wondering if this issue is related to vehicals randomly stopping, or trains switching at nodes that aren't supposed to be switches
!  I've tried all the tricks to get it to reconnect, but it just refuses to reconnect.  I've already built this line once.  I'm not keen on rebuilding it a 2nd time.  I've never had issues with TMPE (but, that doesn't necessarily mean it isn't the culprit).  I have, however, read that MOM (Metro Overhaul Mod) can cause issues.  I'm not sure I want to remove/disable the mod for fear I won't be able to work on my map (even if i enable "ghost mode").

attached is a pic of my unfixable broken bus line. there are no lane connectors, and buses are allowed to ignore lane arrows.[cities skylines - broken bus line](https://user-images.githubusercontent.com/31520821/172421951-534aa5e3-7177-49b1-a7f3-b948c7e308fe.png)
.
